File: flutter/lib/core/api/pro_api.dart
import 'package:dio/dio.dart';

import '../../data/api/models/api_error.dart';

class ProApiResult<T> {
  const ProApiResult._({this.data, this.error});

  final T? data;
  final ProApiFailure? error;

  bool get ok => error == null;

  static ProApiResult<T> success<T>(T data) => ProApiResult<T>._(data: data);
  static ProApiResult<T> failure<T>(ProApiFailure error) => ProApiResult<T>._(error: error);
}

class ProApiFailure {
  const ProApiFailure({required this.kind, required this.code, required this.message, this.status});

  final String kind;
  final String code;
  final String message;
  final int? status;
}

class ProApi {
  ProApi(this._dio);

  final Dio _dio;

  ProApiFailure _mapError(Object error) {
    if (error is DioException) {
      final status = error.response?.statusCode;
      final payloadError = error.error;
      if (payloadError is ApiError) {
        final kind = switch (status) {
          401 => 'unauthorized',
          403 => 'forbidden',
          422 => 'validation',
          429 => 'rate_limited',
          int s when s >= 500 => 'server',
          _ => 'unknown'
        };
        return ProApiFailure(kind: kind, code: payloadError.code, message: payloadError.message, status: status);
      }
      return ProApiFailure(kind: 'network', code: 'network_error', message: error.message ?? 'Network error', status: status);
    }
    return const ProApiFailure(kind: 'unknown', code: 'unknown_error', message: 'Unknown error');
  }

  Future<ProApiResult<T>> _wrap<T>(Future<T> Function() fn) async {
    try {
      final data = await fn();
      return ProApiResult.success(data);
    } catch (error) {
      return ProApiResult.failure(_mapError(error));
    }
  }

  Future<ProApiResult<Map<String, dynamic>>> me() => _wrap(() async => (await _dio.get<Map<String, dynamic>>('/me')).data ?? {});
  Future<ProApiResult<void>> logout(String refreshToken) => _wrap(() async => _dio.post('/auth/logout', data: {'refresh_token': refreshToken, 'revoke_family': true}));
  Future<ProApiResult<Map<String, dynamic>>> refresh(String refreshToken) => _wrap(() async => (await _dio.post<Map<String, dynamic>>('/auth/refresh', data: {'refresh_token': refreshToken})).data ?? {});
  Future<ProApiResult<void>> requestVerifyEmail(String email) => _wrap(() async => _dio.post('/auth/verify-email/request', data: {'email': email}));
  Future<ProApiResult<void>> confirmVerifyEmail(String code) => _wrap(() async => _dio.post('/auth/verify-email/confirm', data: {'code': code}));

  Future<ProApiResult<Map<String, dynamic>>> getMyProProfile() => _wrap(() async => (await _dio.get<Map<String, dynamic>>('/pro/me/profile')).data ?? {});
  Future<ProApiResult<Map<String, dynamic>>> updateMyProProfile(Map<String, dynamic> payload) => _wrap(() async => (await _dio.put<Map<String, dynamic>>('/pro/me/profile', data: payload)).data ?? {});
  Future<ProApiResult<Map<String, dynamic>>> getOnboarding() => _wrap(() async => (await _dio.get<Map<String, dynamic>>('/pro/onboarding')).data ?? {});
  Future<ProApiResult<Map<String, dynamic>>> getOnboardingChecks() => _wrap(() async => (await _dio.get<Map<String, dynamic>>('/pro/onboarding/checks')).data ?? {});
  Future<ProApiResult<Map<String, dynamic>>> onboardingStart() => _wrap(() async => (await _dio.post<Map<String, dynamic>>('/pro/onboarding/start')).data ?? {});
  Future<ProApiResult<Map<String, dynamic>>> onboardingCompleteProfile(Map<String, dynamic> payload) => _wrap(() async => (await _dio.post<Map<String, dynamic>>('/pro/onboarding/complete-profile', data: payload)).data ?? {});
  Future<ProApiResult<Map<String, dynamic>>> onboardingSelectNiches(Map<String, dynamic> payload) => _wrap(() async => (await _dio.post<Map<String, dynamic>>('/pro/onboarding/select-niches', data: payload)).data ?? {});
  Future<ProApiResult<Map<String, dynamic>>> onboardingConfigurePackages(Map<String, dynamic> payload) => _wrap(() async => (await _dio.post<Map<String, dynamic>>('/pro/onboarding/configure-packages', data: payload)).data ?? {});
  Future<ProApiResult<Map<String, dynamic>>> onboardingUploadPortfolio(Map<String, dynamic> payload) => _wrap(() async => (await _dio.post<Map<String, dynamic>>('/pro/onboarding/upload-portfolio', data: payload)).data ?? {});
  Future<ProApiResult<Map<String, dynamic>>> onboardingSubmitKyc(Map<String, dynamic> payload) => _wrap(() async => (await _dio.post<Map<String, dynamic>>('/pro/onboarding/submit-kyc', data: payload)).data ?? {});
  Future<ProApiResult<List<dynamic>>> listNiches() => _wrap(() async => (await _dio.get<List<dynamic>>('/niches')).data ?? <dynamic>[]);
  Future<ProApiResult<Map<String, dynamic>>> getMyNiches() => _wrap(() async => (await _dio.get<Map<String, dynamic>>('/pro/niches/mine')).data ?? {});
  Future<ProApiResult<Map<String, dynamic>>> putMyNiches(Map<String, dynamic> payload) => _wrap(() async => (await _dio.put<Map<String, dynamic>>('/pro/niches/mine', data: payload)).data ?? {});
  Future<ProApiResult<Map<String, dynamic>>> createPackage(Map<String, dynamic> payload) => _wrap(() async => (await _dio.post<Map<String, dynamic>>('/pro/me/packages', data: payload)).data ?? {});
  Future<ProApiResult<Map<String, dynamic>>> updatePackage(String packageId, Map<String, dynamic> payload) => _wrap(() async => (await _dio.put<Map<String, dynamic>>('/pro/me/packages/$packageId', data: payload)).data ?? {});
  Future<ProApiResult<Map<String, dynamic>>> disablePackage(String packageId) => _wrap(() async => (await _dio.post<Map<String, dynamic>>('/pro/me/packages/$packageId/disable')).data ?? {});
  Future<ProApiResult<Map<String, dynamic>>> tagPortfolioMediaNiches(String mediaAssetId, Map<String, dynamic> payload) => _wrap(() async => (await _dio.post<Map<String, dynamic>>('/pro/me/portfolio/$mediaAssetId/niches', data: payload)).data ?? {});

  Future<ProApiResult<Map<String, dynamic>>> searchPros(Map<String, dynamic> params) => _wrap(() async => (await _dio.get<Map<String, dynamic>>('/search/pros', queryParameters: params)).data ?? {});
  Future<ProApiResult<Map<String, dynamic>>> getPublicProProfile(String proUserId) => _wrap(() async => (await _dio.get<Map<String, dynamic>>('/pros/$proUserId/public')).data ?? {});

  Future<ProApiResult<Map<String, dynamic>>> getAvailabilityRules() => _wrap(() async => (await _dio.get<Map<String, dynamic>>('/pro/scheduling/availability-rules')).data ?? {});
  Future<ProApiResult<Map<String, dynamic>>> putAvailabilityRules(Map<String, dynamic> payload) => _wrap(() async => (await _dio.put<Map<String, dynamic>>('/pro/scheduling/availability-rules', data: payload)).data ?? {});
  Future<ProApiResult<Map<String, dynamic>>> getSchedulingExceptions() => _wrap(() async => (await _dio.get<Map<String, dynamic>>('/pro/scheduling/exceptions')).data ?? {});
  Future<ProApiResult<Map<String, dynamic>>> putSchedulingExceptions(Map<String, dynamic> payload) => _wrap(() async => (await _dio.put<Map<String, dynamic>>('/pro/scheduling/exceptions', data: payload)).data ?? {});
  Future<ProApiResult<Map<String, dynamic>>> getSchedulingPolicy() => _wrap(() async => (await _dio.get<Map<String, dynamic>>('/pro/scheduling/policy')).data ?? {});
  Future<ProApiResult<Map<String, dynamic>>> putSchedulingPolicy(Map<String, dynamic> payload) => _wrap(() async => (await _dio.put<Map<String, dynamic>>('/pro/scheduling/policy', data: payload)).data ?? {});
  Future<ProApiResult<Map<String, dynamic>>> getCandidateSlots(Map<String, dynamic> params) => _wrap(() async => (await _dio.get<Map<String, dynamic>>('/pro/scheduling/slots', queryParameters: params)).data ?? {});
  Future<ProApiResult<Map<String, dynamic>>> getPublicAvailability(String proUserId, Map<String, dynamic> params) => _wrap(() async => (await _dio.get<Map<String, dynamic>>('/pro/$proUserId/availability', queryParameters: params)).data ?? {});

  Future<ProApiResult<Map<String, dynamic>>> getBookingRequest(String requestId) => _wrap(() async => (await _dio.get<Map<String, dynamic>>('/booking-requests/$requestId')).data ?? {});
  Future<ProApiResult<Map<String, dynamic>>> acceptBookingRequest(String requestId) => _wrap(() async => (await _dio.post<Map<String, dynamic>>('/booking-requests/$requestId/accept')).data ?? {});
  Future<ProApiResult<Map<String, dynamic>>> declineBookingRequest(String requestId, {String? reason}) => _wrap(() async => (await _dio.post<Map<String, dynamic>>('/booking-requests/$requestId/decline', data: reason == null ? {} : {'reason': reason})).data ?? {});
  Future<ProApiResult<Map<String, dynamic>>> cancelBookingRequest(String requestId) => _wrap(() async => (await _dio.post<Map<String, dynamic>>('/booking-requests/$requestId/cancel')).data ?? {});
  Future<ProApiResult<Map<String, dynamic>>> confirmSlot(String bookingRequestId, Map<String, dynamic> payload) => _wrap(() async => (await _dio.post<Map<String, dynamic>>('/pro/bookings/$bookingRequestId/confirm-slot', data: payload)).data ?? {});

  Future<ProApiResult<Map<String, dynamic>>> createGig(Map<String, dynamic> payload) => _wrap(() async => (await _dio.post<Map<String, dynamic>>('/gigs', data: payload)).data ?? {});
  Future<ProApiResult<Map<String, dynamic>>> getGig(String gigId) => _wrap(() async => (await _dio.get<Map<String, dynamic>>('/gigs/$gigId')).data ?? {});
  Future<ProApiResult<Map<String, dynamic>>> cancelGigSlot(String gigId, Map<String, dynamic> payload) => _wrap(() async => (await _dio.post<Map<String, dynamic>>('/gigs/$gigId/cancel-slot', data: payload)).data ?? {});
  Future<ProApiResult<Map<String, dynamic>>> requestReschedule(String gigId, Map<String, dynamic> payload) => _wrap(() async => (await _dio.post<Map<String, dynamic>>('/gigs/$gigId/reschedule-request', data: payload)).data ?? {});
  Future<ProApiResult<Map<String, dynamic>>> getGigConsent(String gigId) => _wrap(() async => (await _dio.get<Map<String, dynamic>>('/gigs/$gigId/consent')).data ?? {});
  Future<ProApiResult<Map<String, dynamic>>> putGigConsent(String gigId, Map<String, dynamic> payload) => _wrap(() async => (await _dio.put<Map<String, dynamic>>('/gigs/$gigId/consent', data: payload)).data ?? {});
  Future<ProApiResult<Map<String, dynamic>>> listGigMedia(String gigId) => _wrap(() async => (await _dio.get<Map<String, dynamic>>('/gigs/$gigId/media')).data ?? {});
  Future<ProApiResult<Map<String, dynamic>>> getGigMediaSignedUrl(String gigId, String mediaAssetId) => _wrap(() async => (await _dio.get<Map<String, dynamic>>('/gigs/$gigId/media/$mediaAssetId/signed-url')).data ?? {});
  Future<ProApiResult<Map<String, dynamic>>> downloadGigMedia(String gigId, String mediaAssetId) => _wrap(() async => (await _dio.get<Map<String, dynamic>>('/gigs/$gigId/media/$mediaAssetId/download')).data ?? {});
  Future<ProApiResult<Map<String, dynamic>>> createGigShareLink(String gigId, Map<String, dynamic> payload) => _wrap(() async => (await _dio.post<Map<String, dynamic>>('/gigs/$gigId/share-links', data: payload)).data ?? {});

  Future<ProApiResult<Map<String, dynamic>>> createProofGalleryForGig(String gigId, Map<String, dynamic> payload) => _wrap(() async => (await _dio.post<Map<String, dynamic>>('/gigs/$gigId/proof-gallery', data: payload)).data ?? {});
  Future<ProApiResult<Map<String, dynamic>>> getProofGallery(String galleryId) => _wrap(() async => (await _dio.get<Map<String, dynamic>>('/proof-galleries/$galleryId')).data ?? {});
  Future<ProApiResult<Map<String, dynamic>>> addProofGalleryItems(String galleryId, Map<String, dynamic> payload) => _wrap(() async => (await _dio.post<Map<String, dynamic>>('/proof-galleries/$galleryId/items', data: payload)).data ?? {});
  Future<ProApiResult<Map<String, dynamic>>> publishProofGallery(String galleryId) => _wrap(() async => (await _dio.post<Map<String, dynamic>>('/proof-galleries/$galleryId/publish')).data ?? {});
  Future<ProApiResult<Map<String, dynamic>>> saveGallerySelection(String galleryId, Map<String, dynamic> payload) => _wrap(() async => (await _dio.post<Map<String, dynamic>>('/proof-galleries/$galleryId/selections', data: payload)).data ?? {});
  Future<ProApiResult<Map<String, dynamic>>> submitGallerySelection(String galleryId, Map<String, dynamic> payload) => _wrap(() async => (await _dio.post<Map<String, dynamic>>('/proof-galleries/$galleryId/selections/submit', data: payload)).data ?? {});
  Future<ProApiResult<Map<String, dynamic>>> createUpsellIntent(String galleryId, Map<String, dynamic> payload) => _wrap(() async => (await _dio.post<Map<String, dynamic>>('/proof-galleries/$galleryId/upsell/create-intent', data: payload)).data ?? {});
  Future<ProApiResult<Map<String, dynamic>>> getProofGalleryDownloads(String galleryId) => _wrap(() async => (await _dio.get<Map<String, dynamic>>('/proof-galleries/$galleryId/downloads')).data ?? {});

  Future<ProApiResult<Map<String, dynamic>>> createPhotoUpload(Map<String, dynamic> payload) => _wrap(() async => (await _dio.post<Map<String, dynamic>>('/media/photos/uploads', data: payload)).data ?? {});
  Future<ProApiResult<Map<String, dynamic>>> completePhotoUpload(String mediaAssetId, Map<String, dynamic> payload) => _wrap(() async => (await _dio.post<Map<String, dynamic>>('/media/photos/$mediaAssetId/complete', data: payload)).data ?? {});
  Future<ProApiResult<Map<String, dynamic>>> getMediaAsset(String mediaAssetId) => _wrap(() async => (await _dio.get<Map<String, dynamic>>('/media/$mediaAssetId')).data ?? {});

  Future<ProApiResult<Map<String, dynamic>>> listProThreads(Map<String, dynamic> params) => _wrap(() async => (await _dio.get<Map<String, dynamic>>('/pro/chat/threads', queryParameters: params)).data ?? {});
  Future<ProApiResult<Map<String, dynamic>>> getProThread(String threadId) => _wrap(() async => (await _dio.get<Map<String, dynamic>>('/pro/chat/threads/$threadId')).data ?? {});
  Future<ProApiResult<Map<String, dynamic>>> sendProMessage(String threadId, Map<String, dynamic> payload) => _wrap(() async => (await _dio.post<Map<String, dynamic>>('/pro/chat/threads/$threadId/messages', data: payload)).data ?? {});
  Future<ProApiResult<Map<String, dynamic>>> getAIDraft(String threadId, Map<String, dynamic> payload) => _wrap(() async => (await _dio.post<Map<String, dynamic>>('/pro/chat/threads/$threadId/ai-draft', data: payload)).data ?? {});

  Future<ProApiResult<Map<String, dynamic>>> getEarningsBalance() => _wrap(() async => (await _dio.get<Map<String, dynamic>>('/pro/earnings/balance')).data ?? {});
  Future<ProApiResult<Map<String, dynamic>>> getEarningsLedger(Map<String, dynamic> params) => _wrap(() async => (await _dio.get<Map<String, dynamic>>('/pro/earnings/ledger', queryParameters: params)).data ?? {});
  Future<ProApiResult<Map<String, dynamic>>> getPayouts(Map<String, dynamic> params) => _wrap(() async => (await _dio.get<Map<String, dynamic>>('/pro/payouts', queryParameters: params)).data ?? {});
  Future<ProApiResult<Map<String, dynamic>>> getPayoutAccount() => _wrap(() async => (await _dio.get<Map<String, dynamic>>('/pro/payouts/account')).data ?? {});
  Future<ProApiResult<Map<String, dynamic>>> putPayoutAccount(Map<String, dynamic> payload) => _wrap(() async => (await _dio.put<Map<String, dynamic>>('/pro/payouts/account', data: payload)).data ?? {});
  Future<ProApiResult<Map<String, dynamic>>> requestPayout(Map<String, dynamic> payload) => _wrap(() async => (await _dio.post<Map<String, dynamic>>('/pro/payouts/request', data: payload)).data ?? {});

  Future<ProApiResult<Map<String, dynamic>>> track(String name, Map<String, dynamic> props) => _wrap(() async => (await _dio.post<Map<String, dynamic>>('/analytics', data: {'name': name, 'props': props})).data ?? {});
}

File: web/lib/api/proApi.ts
"use client";

import { ApiError, apiRequest } from "@/lib/api/client";

export type ApiErrorKind = "unauthorized" | "forbidden" | "validation" | "rate_limited" | "server" | "network" | "unknown";

export type ProApiError = {
  kind: ApiErrorKind;
  code: string;
  message: string;
  status?: number;
  details?: Record<string, unknown>;
};

export type Result<T> = { ok: true; data: T } | { ok: false; error: ProApiError };

function toError(error: unknown): ProApiError {
  if (error instanceof ApiError) {
    const status = Number(error.details?.status || 0) || undefined;
    const code = error.code || "unknown";
    const kind: ApiErrorKind =
      code === "unauthorized" || status === 401
        ? "unauthorized"
        : code === "forbidden" || status === 403
          ? "forbidden"
          : code === "validation_error" || status === 422
            ? "validation"
            : status === 429
              ? "rate_limited"
              : status && status >= 500
                ? "server"
                : "unknown";
    return { kind, code, message: error.message, status, details: error.details };
  }
  return { kind: "network", code: "network_error", message: error instanceof Error ? error.message : "Network error" };
}

async function wrap<T>(fn: () => Promise<T>): Promise<Result<T>> {
  try {
    const data = await fn();
    return { ok: true, data };
  } catch (error) {
    return { ok: false, error: toError(error) };
  }
}

function get<T>(path: string, accessToken?: string | null, query?: Record<string, string | number | boolean | undefined>) {
  return apiRequest<T>(path, { accessToken, query });
}
function post<T>(path: string, body: unknown, accessToken?: string | null) {
  return apiRequest<T>(path, { method: "POST", accessToken, body: JSON.stringify(body) });
}
function put<T>(path: string, body: unknown, accessToken?: string | null) {
  return apiRequest<T>(path, { method: "PUT", accessToken, body: JSON.stringify(body) });
}

export const proApi = {
  me: (accessToken?: string | null) => wrap(() => get<any>("/me", accessToken)),
  logout: (refreshToken: string | null, accessToken?: string | null) => wrap(() => post<void>("/auth/logout", { refresh_token: refreshToken, revoke_family: true }, accessToken)),
  refresh: (refreshToken: string) => wrap(() => post<{ access_token: string; refresh_token: string; expires_in: number }>("/auth/refresh", { refresh_token: refreshToken })),
  requestVerifyEmail: (email: string) => wrap(() => post<void>("/auth/verify-email/request", { email })),
  confirmVerifyEmail: (code: string) => wrap(() => post<void>("/auth/verify-email/confirm", { code })),

  getMyProProfile: (accessToken?: string | null) => wrap(() => get<any>("/pro/me/profile", accessToken)),
  updateMyProProfile: (payload: Record<string, unknown>, accessToken?: string | null) => wrap(() => put<any>("/pro/me/profile", payload, accessToken)),
  getOnboarding: (accessToken?: string | null) => wrap(() => get<any>("/pro/onboarding", accessToken)),
  getOnboardingChecks: (accessToken?: string | null) => wrap(() => get<any>("/pro/onboarding/checks", accessToken)),
  onboardingStart: (accessToken?: string | null) => wrap(() => post<any>("/pro/onboarding/start", {}, accessToken)),
  onboardingCompleteProfile: (payload: Record<string, unknown>, accessToken?: string | null) => wrap(() => post<any>("/pro/onboarding/complete-profile", payload, accessToken)),
  onboardingSelectNiches: (payload: Record<string, unknown>, accessToken?: string | null) => wrap(() => post<any>("/pro/onboarding/select-niches", payload, accessToken)),
  onboardingConfigurePackages: (payload: Record<string, unknown>, accessToken?: string | null) => wrap(() => post<any>("/pro/onboarding/configure-packages", payload, accessToken)),
  onboardingUploadPortfolio: (payload: Record<string, unknown>, accessToken?: string | null) => wrap(() => post<any>("/pro/onboarding/upload-portfolio", payload, accessToken)),
  onboardingSubmitKyc: (payload: Record<string, unknown>, accessToken?: string | null) => wrap(() => post<any>("/pro/onboarding/submit-kyc", payload, accessToken)),
  listNiches: (accessToken?: string | null) => wrap(() => get<any>("/niches", accessToken)),
  getMyNiches: (accessToken?: string | null) => wrap(() => get<any>("/pro/niches/mine", accessToken)),
  putMyNiches: (payload: Record<string, unknown>, accessToken?: string | null) => wrap(() => put<any>("/pro/niches/mine", payload, accessToken)),
  createPackage: (payload: Record<string, unknown>, accessToken?: string | null) => wrap(() => post<any>("/pro/me/packages", payload, accessToken)),
  updatePackage: (packageId: string, payload: Record<string, unknown>, accessToken?: string | null) => wrap(() => put<any>(`/pro/me/packages/${packageId}`, payload, accessToken)),
  disablePackage: (packageId: string, accessToken?: string | null) => wrap(() => post<any>(`/pro/me/packages/${packageId}/disable`, {}, accessToken)),
  tagPortfolioMediaNiches: (mediaAssetId: string, payload: Record<string, unknown>, accessToken?: string | null) => wrap(() => post<any>(`/pro/me/portfolio/${mediaAssetId}/niches`, payload, accessToken)),

  searchPros: (params: Record<string, string | number | boolean | undefined>, accessToken?: string | null) => wrap(() => get<any>("/search/pros", accessToken, params)),
  getPublicProProfile: (proUserId: string, accessToken?: string | null) => wrap(() => get<any>(`/pros/${proUserId}/public`, accessToken)),

  getAvailabilityRules: (accessToken?: string | null) => wrap(() => get<any>("/pro/scheduling/availability-rules", accessToken)),
  putAvailabilityRules: (payload: Record<string, unknown>, accessToken?: string | null) => wrap(() => put<any>("/pro/scheduling/availability-rules", payload, accessToken)),
  getSchedulingExceptions: (accessToken?: string | null) => wrap(() => get<any>("/pro/scheduling/exceptions", accessToken)),
  putSchedulingExceptions: (payload: Record<string, unknown>, accessToken?: string | null) => wrap(() => put<any>("/pro/scheduling/exceptions", payload, accessToken)),
  getSchedulingPolicy: (accessToken?: string | null) => wrap(() => get<any>("/pro/scheduling/policy", accessToken)),
  putSchedulingPolicy: (payload: Record<string, unknown>, accessToken?: string | null) => wrap(() => put<any>("/pro/scheduling/policy", payload, accessToken)),
  getCandidateSlots: (params: Record<string, string | number | boolean | undefined>, accessToken?: string | null) => wrap(() => get<any>("/pro/scheduling/slots", accessToken, params)),
  getPublicAvailability: (proUserId: string, params: Record<string, string | number | boolean | undefined>, accessToken?: string | null) => wrap(() => get<any>(`/pro/${proUserId}/availability`, accessToken, params)),

  getBookingRequest: (requestId: string, accessToken?: string | null) => wrap(() => get<any>(`/booking-requests/${requestId}`, accessToken)),
  acceptBookingRequest: (requestId: string, accessToken?: string | null) => wrap(() => post<any>(`/booking-requests/${requestId}/accept`, {}, accessToken)),
  declineBookingRequest: (requestId: string, reason: string | null, accessToken?: string | null) => wrap(() => post<any>(`/booking-requests/${requestId}/decline`, reason ? { reason } : {}, accessToken)),
  cancelBookingRequest: (requestId: string, accessToken?: string | null) => wrap(() => post<any>(`/booking-requests/${requestId}/cancel`, {}, accessToken)),
  confirmSlot: (bookingRequestId: string, payload: Record<string, unknown>, accessToken?: string | null) => wrap(() => post<any>(`/pro/bookings/${bookingRequestId}/confirm-slot`, payload, accessToken)),

  getGig: (gigId: string, accessToken?: string | null) => wrap(() => get<any>(`/gigs/${gigId}`, accessToken)),
  createGig: (payload: Record<string, unknown>, accessToken?: string | null) => wrap(() => post<any>("/gigs", payload, accessToken)),
  cancelGigSlot: (gigId: string, payload: Record<string, unknown>, accessToken?: string | null) => wrap(() => post<any>(`/gigs/${gigId}/cancel-slot`, payload, accessToken)),
  requestReschedule: (gigId: string, payload: Record<string, unknown>, accessToken?: string | null) => wrap(() => post<any>(`/gigs/${gigId}/reschedule-request`, payload, accessToken)),
  getGigConsent: (gigId: string, accessToken?: string | null) => wrap(() => get<any>(`/gigs/${gigId}/consent`, accessToken)),
  putGigConsent: (gigId: string, payload: Record<string, unknown>, accessToken?: string | null) => wrap(() => put<any>(`/gigs/${gigId}/consent`, payload, accessToken)),
  listGigMedia: (gigId: string, accessToken?: string | null) => wrap(() => get<any>(`/gigs/${gigId}/media`, accessToken)),
  getGigMediaSignedUrl: (gigId: string, mediaAssetId: string, accessToken?: string | null) => wrap(() => get<any>(`/gigs/${gigId}/media/${mediaAssetId}/signed-url`, accessToken)),
  downloadGigMedia: (gigId: string, mediaAssetId: string, accessToken?: string | null) => wrap(() => get<any>(`/gigs/${gigId}/media/${mediaAssetId}/download`, accessToken)),
  createGigShareLink: (gigId: string, payload: Record<string, unknown>, accessToken?: string | null) => wrap(() => post<any>(`/gigs/${gigId}/share-links`, payload, accessToken)),

  createProofGalleryForGig: (gigId: string, payload: Record<string, unknown>, accessToken?: string | null) => wrap(() => post<any>(`/gigs/${gigId}/proof-gallery`, payload, accessToken)),
  getProofGallery: (galleryId: string, accessToken?: string | null) => wrap(() => get<any>(`/proof-galleries/${galleryId}`, accessToken)),
  addProofGalleryItems: (galleryId: string, payload: Record<string, unknown>, accessToken?: string | null) => wrap(() => post<any>(`/proof-galleries/${galleryId}/items`, payload, accessToken)),
  publishProofGallery: (galleryId: string, accessToken?: string | null) => wrap(() => post<any>(`/proof-galleries/${galleryId}/publish`, {}, accessToken)),
  saveGallerySelection: (galleryId: string, payload: Record<string, unknown>, accessToken?: string | null) => wrap(() => post<any>(`/proof-galleries/${galleryId}/selections`, payload, accessToken)),
  submitGallerySelection: (galleryId: string, payload: Record<string, unknown>, accessToken?: string | null) => wrap(() => post<any>(`/proof-galleries/${galleryId}/selections/submit`, payload, accessToken)),
  createUpsellIntent: (galleryId: string, payload: Record<string, unknown>, accessToken?: string | null) => wrap(() => post<any>(`/proof-galleries/${galleryId}/upsell/create-intent`, payload, accessToken)),
  getProofGalleryDownloads: (galleryId: string, accessToken?: string | null) => wrap(() => get<any>(`/proof-galleries/${galleryId}/downloads`, accessToken)),

  createPhotoUpload: (payload: Record<string, unknown>, accessToken?: string | null) => wrap(() => post<any>("/media/photos/uploads", payload, accessToken)),
  completePhotoUpload: (mediaAssetId: string, payload: Record<string, unknown>, accessToken?: string | null) => wrap(() => post<any>(`/media/photos/${mediaAssetId}/complete`, payload, accessToken)),
  getMediaAsset: (mediaAssetId: string, accessToken?: string | null) => wrap(() => get<any>(`/media/${mediaAssetId}`, accessToken)),

  listProThreads: (params: Record<string, string | number | boolean | undefined>, accessToken?: string | null) => wrap(() => get<any>("/pro/chat/threads", accessToken, params)),
  getProThread: (threadId: string, accessToken?: string | null) => wrap(() => get<any>(`/pro/chat/threads/${threadId}`, accessToken)),
  sendProMessage: (threadId: string, payload: Record<string, unknown>, accessToken?: string | null) => wrap(() => post<any>(`/pro/chat/threads/${threadId}/messages`, payload, accessToken)),
  getAIDraft: (threadId: string, payload: Record<string, unknown>, accessToken?: string | null) => wrap(() => post<any>(`/pro/chat/threads/${threadId}/ai-draft`, payload, accessToken)),

  getEarningsBalance: (accessToken?: string | null) => wrap(() => get<any>("/pro/earnings/balance", accessToken)),
  getEarningsLedger: (params: Record<string, string | number | boolean | undefined>, accessToken?: string | null) => wrap(() => get<any>("/pro/earnings/ledger", accessToken, params)),
  getPayouts: (params: Record<string, string | number | boolean | undefined>, accessToken?: string | null) => wrap(() => get<any>("/pro/payouts", accessToken, params)),
  getPayoutAccount: (accessToken?: string | null) => wrap(() => get<any>("/pro/payouts/account", accessToken)),
  putPayoutAccount: (payload: Record<string, unknown>, accessToken?: string | null) => wrap(() => put<any>("/pro/payouts/account", payload, accessToken)),
  requestPayout: (payload: Record<string, unknown>, accessToken?: string | null) => wrap(() => post<any>("/pro/payouts/request", payload, accessToken)),

  track: (name: string, props: Record<string, unknown>, accessToken?: string | null) => wrap(() => post<any>("/analytics", { name, props }, accessToken)),
};
